-XUnstable, protecting GHC's unstable features (#524)
ghc-proposals/ghc-proposals: proposals/0000-unstable.rst

in release notes, we recognize it is impractical to expect every GHC user to read these
resources.

This proposal introduces a new extension ``-XUnstable``. Unstable features
will be available only when ``-XUnstable`` is specified. Unstable features might
change without notice, without a migration plan, and without a proposal, though
notice of any changes would appear in release notes. In addition, we might choose
to use this proposal process about unstable features, if collecting community
feedback would be helpful.

Moving
an unstable feature out from the unstable umbrella would require a proposal.

This proposal marks importing any module from ``ghc-prim`` as requiring ``-XUnstable``,
though we expect to expand the set of features affected by ``-XUnstable`` over time.

This proposal also includes a separate extension, ``-XStable``, which prevents
``-XUnstable`` from being used. A project's ``.cabal`` file might then specify
``-XStable`` to prevent any of the project's modules from using ``-XUnstable``.
There is no ``-XNoStable``, unlike most other extensions. This ``-XStable`` component
is optional and may be removed from this proposal without affecting the

> utility of ``-XUnstable``.

Reviewer: I’d leave this out. So far we have no precedence for “.cabal file gets to set boundaries for haskell files”, and it assumes a division of roles with different “privileges” between those editing the .cabal file and those editing source files that I am not sure is realistic. 

Using `-XUnstable` is just one of many many things a local programming policy might want to forbid (such as other extensions, unsafe features, use of emoji keywords…), and is better served by a custom hlint database maybe.

Credit: This idea came up in a conversation between Richard Eisenberg and Simon
Peyton Jones, though Richard fleshed out the details and wrote this proposal.

Motivation
----------

``-XUnstable``
~~~~~~~~~~~~~~

It is frequently the case that the GHC developers know a feature is likely to
change soon (or not-so-soon), and yet it is hard to spread this knowledge to
the wider Haskell community. Examples:

* For years, we GHC developers knew that ``-XImpredicativeTypes`` was more of
  a bug than a feature. It had no specification, and any code that type-checked

>   because of it was simply lucky. The feature was in essence unstable. Since

Reviewer: Did we know that even already when `-XImpredicativeTypes` was fist introduced, or was that hindsight? This only counts as motivation in the former case, does it?

  some years later with ``-XTypeInType``; the latter extension ended up getting
  deprecated, in the end.

These are all examples of features we introduced despite their known instability.
However, we might imagine that, with the ability to guard some features behind
a scary flag like ``-XUnstable``, we would have been even more bold in rolling
out other ideas. In many cases, we gain knowledge over time by having a feature
available; this knowledge can help inform the design of the feature. An excellent
example is ``-XOverloadedRecordDot``. This extension was essentially implemented
within Digital Asset's internal version of Haskell for some time before being proposed in GHC.
Through years of experimentation and user feedback, the design was refined. GHC then
benefited from this history in getting a new, yet battle-tested extension. If Digital
Asset had not invested in this way, though, we would have no way of gaining experience
without trapping ourselves in a potentially backward-incompatible corner. With ``-XUnstable``,
we might be able to roll out more features, learn from their use, and then refine.

``-XStable``
~~~~~~~~~~~~

> Unstable features are, well, unstable. This means that some developers and enterprises

Reviewer 2: ```suggestion
Unstable features are, well, unstable. Consequently, some developers and enterprises
```

"This means that" seems to be ambiguous. I think this is the intended meaning.

even without ``-XStable``), but ``-XStable`` depends on ``-XUnstable``.

``-XUnstable``
~~~~~~~~~~~~~~

Technical specification
#######################

1. Introduce a new extension ``-XUnstable``. This extension is off by default and
   not a part of any extension set.

#. When a module from the ``ghc-prim`` package is imported, if ``-XUnstable`` is not
   specified, report an error informing the user that modules from ``ghc-prim`` are
   unstable. After warning the user of the consequences, the error would inform users
   that specifying ``-XUnstable`` makes the error go away.

#. Remove ``FUN`` from the export list of ``GHC.Exts`` (so that ``FUN`` becomes unstable).

#. (Technically beyond the scope of this proposal process) The module-level documentation
   for each of the exported modules of ``ghc-prim`` will include a note that users
   should not import that module directly; instead, users will be directed where to
   find the imports they need.

#. (Technically beyond the scope of this proposal process) Tooling that helps
   users automatically insert extensions (i.e. HLS) would not automatically enable

>    ``-XUnstable`` from this message; users would have to add it manually.

Reviewer: What is itabout `-XUnstable` that makes it deserve that treatment? Are there other extensions that should not be “easily enabled”? (Maybe `Trustworthy`, `IncoherentInstances`)?

And is it even good style? If the user already added `ghc-prim` to their cabal file, and added the imports, they are very likely determined to use `ghc-prim` and hopefully know why. Why make them do manual busy work that the tool could do for them?

(Yes, it’s not technically part of the proposal and the HLS authors might just choose to ignore it, or implement it by allowing automatic after a warning.)

   is the primitive function type, but its current kind is likely to change.

Costs and Drawbacks
-------------------

1. Some have argued that GHC language extensions should be a way of guarding features
   that have not yet made it into the next language standard. Instead, the ``-XUnstable``
   and ``-XStable`` extensions are more like configuration options than language extensions.
   Accepting this proposal moves us further from the vision that some future version of
   Haskell will subsume all extensions.

#. The naming of these extensions may be confusing. In particular, users may wonder
   what the difference between ``-XNoUnstable`` and ``-XStable`` is, or think that
   (the non-existent) ``-XNoStable`` is the same as ``-XUnstable``.

Alternatives
------------

1. Instead of making a new language extension, we could imagine a new warning ``-Wunstable-features``.
   This warning would be werror-by-default (no warnings are like this today) but could be disabled
   with ``-Wno-unstable-features``. Similarly, ``-XStable`` could become something like ``-fstable``,

>    but this seems harder to specify in a ``.cabal`` file.

Reviewer: This is an interesting point. The `nix` tool guard additional subcommand (a bit like extensions) under an “enable experimental features” configuration option.

So using a flag `-fenable-experimental-features` or `-Wno-unstable-features` is an alternative worth discussing.

I don’t think it’s harder to specify this in the `.cabal` file`, `ghc-options: -Wno-unstable-features`. People already know how to set warning flags in the cabal file. And those who don’t know are certainly not the target audience for this flag.
